metric_learn/rca.py

```diff
@@ -62,32 +79,42 @@ def fit(self, data, chunks):
         when ``chunks[i] == j``, point i belongs to chunklet j.
     """
     data, chunks, num_chunks, d = self._process_inputs(data, chunks)
-
-    # mean center
-    data -= data.mean(axis=0)
-
-    # mean center each chunklet separately
-    chunk_mask = chunks != -1
-    chunk_data = data[chunk_mask]
-    chunk_labels = chunks[chunk_mask]
-    for c in xrange(num_chunks):
-      mask = chunk_labels == c
-      chunk_data[mask] -= chunk_data[mask].mean(axis=0)
-
-    # "inner" covariance of chunk deviations
+    chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
     inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+    rank = np.linalg.matrix_rank(inner_cov)
+
+    # If the inner covariance matrix is not full rank,
+    # the input data are first projected with a PCA to a space of dimension rank.
+    M_pca = np.identity(d)
+    if rank < d:
+      pca = decomposition.PCA(n_components = rank)
+      pca.fit(data)
+      M_pca = pca.components_
+      data = pca.transform(data)
+      chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
+      inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+      rank = np.linalg.matrix_rank(inner_cov)
```

Comment:
Thanks for the meta-analysis, this really helps. I like the idea of starting with a PCA transformation unconditionally. The principal components are sorted by how much variance they explain, so we can just do this once and take the first `d` dimensions that produce a full-rank inner covariance fairly cheaply.

We should still let the user choose an embedding dimension, so the process might look like:

1. If the user asks for `d` dimensions (defaulting to `X.shape[1]`), compute the first `d` principal components.
2. Compute the inner covariance matrix and find its rank, `r`, using the standard non-null check.
3. If `r < d`, search for the appropriate `d` principal components and warn the user.

I think this avoids the need for any additional parameters and keeps the implementation pretty straightforward. If the user wants to prune small-but-non-null singular values, they can do so by setting `d`.